protocol/bc/hash.go

```diff
@@ -1,103 +1,92 @@
 package bc
 
 import (
-	"bytes"
 	"database/sql/driver"
-	"encoding/hex"
-	"encoding/json"
-	"fmt"
+	"encoding/binary"
 	"io"
 
 	"golang.org/x/crypto/sha3"
 
 	"chain/crypto/sha3pool"
 	"chain/encoding/blockchain"
-	"chain/errors"
 )
 
-// Hash represents a 256-bit hash.  By convention, Hash objects are
-// typically passed as values, not as pointers.
-type Hash [32]byte
+// Hash represents a 256-bit hash.
 
-var EmptyStringHash = sha3.Sum256(nil)
+var EmptyStringHash = NewHash(sha3.Sum256(nil))
 
-// String returns the bytes of h encoded in hex.
-func (h Hash) String() string {
-	b, _ := h.MarshalText() // #nosec
-	return string(b)
+func NewHash(b [32]byte) (h Hash) {
+	h.FromByte32(b)
+	return h
+}
+
+func (h Hash) Byte32() (b32 [32]byte) {
+	binary.BigEndian.PutUint64(b32[0:8], h.V0)
+	binary.BigEndian.PutUint64(b32[8:16], h.V1)
+	binary.BigEndian.PutUint64(b32[16:24], h.V2)
+	binary.BigEndian.PutUint64(b32[24:32], h.V3)
+	return b32
+}
+
+func (h *Hash) FromByte32(b32 [32]byte) {
```

Comment:
This is bikeshed territory, but I think this method should be called either Set or Init, for the same reason the constructor isn't (and shouldn't be) NewHashFromByte32, and because it's a side-effect procedure so its name should be a verb saying what it does.

If we had several different types that could initialize a Hash, then it would make sense to call them SetFoo, SetBar, etc. But there is just one. (And even if there were several, the "main" one should still probably be just Set.)